server/src/main/java/org/elasticsearch/common/util/BigArrays.java

```diff
@@ -356,6 +356,88 @@ public T set(long index, T value) {
 
     }
 
+    abstract public static class BinarySearcher<T>{
+
+        /**
+         * @return a negative integer, zero, or a positive integer if the
+         * array's value at <code>index</code> is less than, equal to, or greater than <code>value</code>
+         */
+        abstract protected int compare(int index, T value);
+
+        /**
+         * @return the magnitude of the distance between <code>value</code> and the element at <code>index</code>
+         * It will usually be <code>Math.abs(array[index] - value)</code>
+         */
+        abstract protected double distance(int index, T value);
+
+        /**
+         * @return the index who's underlying value is closest to <code>value</code>
+         */
+        private int getClosestIndex(int index1, int index2, T value){
+            if(distance(index1, value) < distance(index2, value)){
+                return index1;
+            } else {
+                return index2;
+            }
+        }
+
+        /**
+         * Uses a binary search to determine the index of the closest element to <code>value</code> that lies in
+         * the index range {from, ... , to}
+         *
+         * @return the index of the closest element
+         *
+         * Requires: The underlying array should be sorted
+         **/
+        public int search(int from, int to, T value){
```

Comment:
I got rid of the `<T>` and then overrode this method in `DoubleBinarySearcher` so that the interface was easier to use. But I'd be happy to remove the override if you don't think it's a good idea.

Reply:
OOOOH! I hadn't thought about that. I'll take a look.